docs: Config Migration Guide (#17230)
eslint/eslint: docs/src/use/configure/migration-guide.md

Reviewer: The user guide lists [migration instructions](https://github.com/eslint/eslint/blob/ed60afd4450e769a975447178299446f4439d926/docs/src/use/index.md?plain=1#L43-L54) for all major versions up to 8.0.0. I was thinking that in order to maintain the current convention, this file should be named `migrating-to-9.0.0` or `migrate-to-9.0.0`.
I can't say if it would be good to add a link to this new guide now or wait until version 9.0.0 is released.

Reviewer 2: @fasttime this would separate from the major version migration guides as it will span multiple versions.

---
title: Configuration Migration Guide
eleventyNavigation:
    key: migration guide
    parent: configure
    title: Configuration Migration Guide
    order: 8
---

This guide provides an overview of how you can migrate your ESLint configuration file from the `.eslintrc` format to the new configuration file format, `eslint.config.js`.

To learn more about the `eslint.config.js` file format, refer to [this blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/).

For reference information on these configuration formats, refer to the following documentation:

* [`.eslintrc` configuration files](configuration-files)
* [`eslint.config.js` configuration files](configuration-files-new)

## Start Using `eslint.config.js`

As of ESLint v9.0.0, the `eslint.config.js` file format is the default configuration file format. If you are using ESLint v9.0.0 or later, you can start using the `eslint.config.js` file format without any additional configuration.

> To use `eslint.config.js` with ESLint v8.x.x, place a `eslint.config.js` file in the root of your project **or** set an `ESLINT_USE_FLAT_CONFIG` environment variable to `true`.

Author: to reviewer: is there a specific version here where as long as `eslint.config.js` is present (even if there's an .eslintrc` file, the flat config is used by default?

Reviewer: `eslint.config.js` lookup was introduced in [v8.23.0](https://github.com/eslint/eslint/releases/tag/v8.23.0). `ESLINT_USE_FLAT_CONFIG` came later in [v8.26.0](https://github.com/eslint/eslint/releases/tag/v8.26.0). Not sure if we need to go into that detail here, especially since we are assuming that ESLint 9 has been already released.

Author: thanks, refactoring to just 'v8'

```suggestion
To use `eslint.config.js` with ESLint v8, place a `eslint.config.js` file in the root of your project **or** set an `ESLINT_USE_FLAT_CONFIG` environment variable to `true`.
```

> ## Key Differences between Configuration Formats

Author: to reviewer: did i miss anything here?

Reviewer 3: I think the globally-ignoring should be highlighted: https://eslint.org/docs/latest/use/configure/configuration-files-new#globally-ignoring-files-with-ignores

as the old `.eslintignore` will be no longer working.

Reviewer: A few more differences that we may want to point out:

* The `root` option no longer exist.
* The `files` option cannot be a single string anymore, it must be an array.
* Files whose name starts with a dot (e.g. `.dotfile.js`) are no longer ignored by default.
* The `parser` and `parserOptions` options are now located under the new setting `languageOptions`.
* `parser` must be an object, not a string.
* The `noInlineConfig` and `reportUnusedDisableDirectives` options are now located under the new setting `linterOptions`.
* Environments associated with an ECMAScript version (e.g. `es6`, `es2020`) can be replaced with `languageOptions.ecmaVersion`, other environments must be replaced with `globals`.
* It is typically not necessary to set `parserOptions.ecmaVersion` if `languageOptions.ecmaVersion` is set.
* It is typically not necessary to set `parserOptions.sourceType` if `languageOptions.sourceType` is set.
* The `sourceType` option now supports the new value "commonjs" (`.eslintrc` supports it too, but it was never documented).

Author: added a new 'ignoring files' section and an 'additional changes section w this info

   }
];
```

### Glob-Based Configs

#### .eslintrc

By default, .`eslintrc` files lint all files (except those covered by `.gitignore`) in the directory in which they’re placed and child directories:

```javascript
// .eslintrc.js
module.exports = {
    // ...other config
    extends: "eslint:recommended",
};
```

If you want to have different configurations for different file glob patterns, you can specify them in the `overrides` property:

```javascript
// .eslintrc.js
module.exports = {
    // ...other config
    extends: "eslint:recommended",
    overrides: [
        {
            files: ["src/**/*"],
            rules: {
                semi: ["warn", "always"]
            }
        },
        {
            files:["test/**/*"],
            rules: {
                "no-console": "off"
            }
        }
    ]
};
```

#### eslint.config.js

By default, `eslint.config.js` files support different glob pattern-based configs in exported array.

Basically, all configuration in the `eslint.config.js` file is like the .`eslintrc` `overrides` property.

Here is a configuration with the default glob pattern (`**/*.{js,mjs,cjs}`):

```javascript
// eslint.config.js
export default [

>     "eslint:recommended", // Recommended config applied to all files

Reviewer 3: `@eslint/js` should be used instead.

Author: didn't know about this. will refactor!

    // Override the recommended config
    {
        rules: {
           indent: ["error", 2],
            "no-unused-vars": "warn"
        }
        // ...other configuration
    }
];
```

To support multiple configs for different glob patterns:

```javascript
// eslint.config.js
export default [

>     "eslint:recommended", // Recommended config applied to all files

Reviewer 3: same here.

export default [
    {
        languageOptions: {
            globals: {
                ...globals.browser,
                myCustomGlobal: "readonly"
            },
            parserOptions: {
                ecmaVersion: 2022,
                sourceType: "module"
            }
        }
        // ...other config
    }
];
```

### Predefined Configs

#### .eslintrc

To use predefined configs, use the `extends` property:

```javascript
// .eslintrc.js
module.exports = {
    // ...other config
    extends: "eslint:recommended",
    rules: {
        semi: ["warn", "always"]
    },
    // ...other config
}
```

ESLint comes with two predefined configs:

* -`eslint:recommended`: the rules recommended by ESLint
* -`eslint:all`: all rules shipped with ESLint

You can also use the `extends` property to extend a custom config. Custom configs can either be paths to local config files or npm package names:

```javascript
// .eslintrc.js
module.exports = {
    // ...other config
    extends: ["eslint:recommended", "./custom-config.js", "eslint-config-my-config"],
    rules: {
        semi: ["warn", "always"]
    },
    // ...other config
}
```

#### eslint.config.js

> `eslint.config.js` supports the built-in `eslint:recommended` and `eslint:all` configs, locally defined configs and npm package configs.

Reviewer 3: the predefined configs are exported in `@eslint/js` now. https://github.com/eslint/eslint/tree/main/packages/js

You can add each of these configs to the exported array.

You must import the modules for local config files and npm package configs with `eslint.config.js`:

```javascript
// eslint.config.js

import customConfig from "./custom-config.js";
import myConfig from "eslint-config-my-config";

export default [
    "eslint:recommended",
    customConfig,
    myConfig,
    {
        rules: {
            semi: ["warn", "always"]
        },
        // ...other config
    }
];
```

Note that because you are just importing JavaScript modules, you can mutate the config objects before exporting them. For example, you might want to have a certain config object only apply to your test files:

```javascript
// eslint.config.js
import customTestConfig from "./custom-test-config.js";
export default [
    "eslint:recommended",
    {
        ...customTestConfig,
        files: ["**/*.test.js"],
    },
];
```

> ## Things That Haven’t Changed between Configuration File Formats

Author: to reviewer: other notable things which should be highlighted?

Reviewer 2: I think we should put this section first, to let people know "these are the things you don't have to worry about".

I think it's also helpful to point out that `globals` are configured the same way, just in a different location.

---
title: Configuration Migration Guide
eleventyNavigation:
    key: migration guide
    parent: configure
    title: Configuration Migration Guide
    order: 8
---

This guide provides an overview of how you can migrate your ESLint configuration file from the `.eslintrc` format to the new configuration file format, `eslint.config.js`.

To learn more about the `eslint.config.js` file format, refer to [this blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/).

For reference information on these configuration formats, refer to the following documentation:

* [`.eslintrc` configuration files](configuration-files)
* [`eslint.config.js` configuration files](configuration-files-new)

## Start Using `eslint.config.js`

> As of ESLint v9.0.0, the `eslint.config.js` file format is the default configuration file format. If you are using ESLint v9.0.0 or later, you can start using the `eslint.config.js` file format without any additional configuration.

Author: I believe the intention is for this guide to be published when v9 comes out. Please correct me if I'm wrong, and I can adjust the text.

> As of ESLint v9.0.0, the `eslint.config.js` file format is the default configuration file format. If you are using ESLint v9.0.0 or later, you can start using the `eslint.config.js` file format without any additional configuration.

Reviewer 2: We'll release this prior to v9 to encourage people to start the transition.

```javascript
// eslint.config.js

import js from "@eslint/js";
import customConfig from "./custom-config.js";
import myConfig from "eslint-config-my-config";

export default [
   js.configs.recommended,
    customConfig,
    myConfig,
    {
        rules: {
            semi: ["warn", "always"]
        },
        // ...other config
    }
];
```

Note that because you are just importing JavaScript modules, you can mutate the config objects before exporting them. For example, you might want to have a certain config object only apply to your test files:

```javascript
// eslint.config.js

import js from "@eslint/js";
import customTestConfig from "./custom-test-config.js";

export default [
    js.configs.recommended,
    {
        ...customTestConfig,
        files: ["**/*.test.js"],
    },
];
```

## Things That Haven’t Changed between Configuration File Formats

While all the above mentioned features have changed from `.eslintrc` to `eslint.config.js` configurations, the following has stayed the same:

* Syntax for adding rules
* Processors
* All functionality. Just the way to configure it has changed.

> * The CLI

Reviewer: `--no-eslintrc` has been replaced with `--no-config-lookup`.

> * The CLI

Author: added new CLI changes section

To use flat config with ESLint v8, place a `eslint.config.js` file in the root of your project **or** set the `ESLINT_USE_FLAT_CONFIG` environment variable to `true`.

## Things That Haven’t Changed between Configuration File Formats

While the configuration file format has changed from eslintrc to flat config, the following has stayed the same:

* Syntax for configuring rules
* Syntax for configuring processors
* The CLI, except for the flag changes noted in [CLI Flag Changes](#cli-flag-changes).
* Global variables are configured the same way, but on a different property (see [Configuring Language Options](#configuring-language-options)).

## Key Differences between Configuration Formats

A few of the most notable differences between the eslintrc and flat config formats are the following:

### Importing Plugins and Custom Parsers

Eslintrc files use string-based import system inside the `plugins` property to load plugins and inside the `extends` property to load external configurations.

> Flat config files use CommonJS `require()` or ES module `import` statements to load plugins and custom parsers.

Author: ```suggestion
Flat config files represent plugins and parsers as JavaScript objects. This means you can use CommonJS `require()` or ES module `import` statements to load plugins and custom parsers from external files.
```

```javascript
// eslint.config.js

import js from "@eslint/js";
import customTestConfig from "./custom-test-config.js";

export default [
    js.configs.recommended,
    {
        ...customTestConfig,
        files: ["**/*.test.js"],
    },
];
```

### Ignoring Files

With eslintrc, you can make ESLint ignore files by creating a separate `.eslintignore` file in the root of your project. The `.eslintignore` file uses the same glob pattern syntax as `.gitignore` files. Alternatively, you can use an `ignorePatterns` property in your eslintrc file.

To ignore files with flat config, you can use the `ignores` property in a config object. The `ignores` property accepts an array of glob patterns. Note that flat config glob patterns do _not_ match dot files (e.g. `.dotfile.js`).

For example, here's a `.eslintignore` example you can use with an eslintrc config:

```shell
# .eslintignore
temp.js
.config/*
# ...other ignored files
```

`ignorePatterns` example:

```javascript
// .eslintrc.js

> {

Reviewer: ```suggestion
module.exports = {
```

    // ...other config
    ignorePatterns: ["temp.js", ".config/*"],

> }

Reviewer: ```suggestion
};
```

* [eslintrc configuration files](configuration-files)
* [flat configuration files](configuration-files-new)

## Start Using Flat Config Files

Starting with ESLint v9.0.0, the flat config file format will be the default configuration file format. Once ESLint v9.0.0 is released, you can start using the flat config file format without any additional configuration.

To use flat config with ESLint v8, place a `eslint.config.js` file in the root of your project **or** set the `ESLINT_USE_FLAT_CONFIG` environment variable to `true`.

## Things That Haven’t Changed between Configuration File Formats

While the configuration file format has changed from eslintrc to flat config, the following has stayed the same:

* Syntax for configuring rules
* Syntax for configuring processors
* The CLI, except for the flag changes noted in [CLI Flag Changes](#cli-flag-changes).
* Global variables are configured the same way, but on a different property (see [Configuring Language Options](#configuring-language-options)).

## Key Differences between Configuration Formats

A few of the most notable differences between the eslintrc and flat config formats are the following:

> ### Importing Plugins and Custom Parsers

Reviewer: To me, the key difference is that in eslintrc files, plugins are specified by a string and loaded from a package by ESLint itself. On the other hand, with the flat config, it's the config file that is responsible for loading a plugin as an object (with `require()` or `import()`), and it's this object that is passed to ESLint.
Ultimately, with the flat config, the plugin doesn't have to be loaded from a package; it can be created by any programmatic means.

Author: rephrased here [ff8aa3a](https://github.com/eslint/eslint/pull/17230/commits/ff8aa3a2de96f3a1cdd7a896ced93b7c69962f55)

}
```

In flat config, you would do the same thing like this:

```javascript
// eslint.config.js

import jsdoc from "eslint-plugin-jsdoc";

export default [
    {
        files: ["**/*.js"],
        plugins: {
            jsdoc: jsdoc
        },
        rules: {
            "jsdoc/require-description": "error",
            "jsdoc/check-values": "error"
        }
    }
];
```

### Custom Parsers

In eslintrc files, importing a custom parser is similar to importing a plugin: you use a string to specify the name of the parser.

In flat config files, import a custom parser as a module, then assign it to the `languageOptions.parser` property of a configuration object.

For example, this eslintrc config file uses the `@babel/eslint-parser` parser:

```javascript
// .eslintrc.js

{
    // ...other config
    parser: "@babel/eslint-parser",
    // ...other config
}
```

In flat config, you would do the same thing like this:

```javascript
// eslint.config.js

import babelParser from "@babel/eslint-parser";

export default [
   {
       // ...other config
       languageOptions: {
           parser: babelParser
       }
       // ...other config
   }
];
```

### Glob-Based Configs

> By default, eslintrc files lint all files (except those covered by `.gitignore`) in the directory in which they’re placed and its child directories. If you want to have different configurations for different file glob patterns, you can specify them in the `overrides` property.

Author: ```suggestion
By default, eslintrc files lint all files (except those covered by `. eslintignore `) in the directory in which they’re placed and its child directories. If you want to have different configurations for different file glob patterns, you can specify them in the `overrides` property.
```

```javascript
// eslint.config.js

import js from "@eslint/js";

export default [
    js.configs.recommended, // Recommended config applied to all files
    // File-pattern specific overrides
    {
        files: ["src/**/*", "test/**/*"],
        rules: {
                semi: ["warn", "always"]
        }
    },
    {
        files:["test/**/*"],
        rules: {
            "no-console": "off"
        }
    }
    // ...other configurations
];
```

### Configuring Language Options

In eslintrc files, you configure various language options across the `env`, `globals` and `parserOptions` properties. Groups of global variables for specific runtimes (e.g `document` and `window` for browser JavaScript; `process` and `require` for Node.js ) are configured with the `env` property.

In flat config files, the `globals`, and `parserOptions` are consolidated under the `languageOptions` key; the `env` property doesn't exist. Groups of global variables for specific runtimes are imported from the [globals](https://www.npmjs.com/package/globals) npm package and included in the `globals` property. You can use the spread operator (`...`) to import multiple globals at once.

> 

Reviewer 4: ```suggestion
```

To fix lint error.

```
docs/src/use/configure/migration-guide.md: 214: MD012/no-multiple-blanks Multiple consecutive blank lines [Expected: 1; Actual: 2]
```

---
title: Configuration Migration Guide
eleventyNavigation:
    key: migration guide
    parent: configure
    title: Configuration Migration Guide
    order: 8
---

This guide provides an overview of how you can migrate your ESLint configuration file from the eslintrc format (typically configured in `.eslintrc.js` or `.eslintrc.json` files) to the new flat config format (typically configured in an `eslint.config.js` file).

To learn more about the flat config format, refer to [this blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/).

For reference information on these configuration formats, refer to the following documentation:

* [eslintrc configuration files](configuration-files)
* [flat configuration files](configuration-files-new)

## Start Using Flat Config Files

Starting with ESLint v9.0.0, the flat config file format will be the default configuration file format. Once ESLint v9.0.0 is released, you can start using the flat config file format without any additional configuration.

> To use flat config with ESLint v8, place a `eslint.config.js` file in the root of your project **or** set an `ESLINT_USE_FLAT_CONFIG` environment variable to `true`.

Reviewer: ```suggestion
To use flat config with ESLint v8, place a `eslint.config.js` file in the root of your project **or** set the `ESLINT_USE_FLAT_CONFIG` environment variable to `true`.
```

Environment variables are uniquely identified by their names.

## Things That Haven’t Changed between Configuration File Formats

While the configuration file format has changed from eslintrc to flat config, the following has stayed the same:

* Syntax for configuring rules

> * Processors

Author: let's do "syntax for ..." for precision

> * Syntax for configuring rules
> * Processors

Author: ```suggestion
* Syntax for configuring rules
* Syntax for configuring processors
```

> * All functionality of ESLint. Just the way to configure it has changed.

Author: ```suggestion
```

* The CLI, except for the flag changes noted in [CLI Flag Changes](#cli-flag-changes).
* Global variables are configured the same way, but on a different property (see [Configuring Language Options](#configuring-language-options)).

## Key Differences between Configuration Formats

A few of the most notable differences between the eslintrc and flat config formats are the following:

### Importing Plugins and Custom Parsers

> Eslintrc files use string-based import system with the `plugins` property to load plugins and `extends` to load configurations from the plugins.

Reviewer 2: ```suggestion
Eslintrc files use a string-based import system with the `plugins` property to load plugins and with the `extends` to load external configurations.
```

Flat config files use CommonJS `requires()` or ESModule `import` to load plugins and custom parsers.

> #### eslintrc Example

Reviewer 2: I don't think we need these headings. Maybe just a sentence like this:

```suggestion
For example, this eslintrc config file loads `eslint-plugin-jsdoc` and configures rules from that plugin:
```

```javascript
// .eslintrc.js

{
    // ...other config
    plugins: ["jsdoc"],
    rules: {
        "jsdoc/require-description": "error",
        "jsdoc/check-values": "error"
    }
    // ...other config
}
```

> #### Flat Config Example

Reviewer 2: ```suggestion
In flat config,  you would do the same thing like this:
```

```javascript
// eslint.config.js

import jsdoc from "eslint-plugin-jsdoc";

export default [
    {
        files: ["**/*.js"],
        plugins: {
            jsdoc: jsdoc
        },
        rules: {
            "jsdoc/require-description": "error",
            "jsdoc/check-values": "error"
        }
    }
];
```

### Custom Parsers

> In eslintrc files, importing a custom parser is very similar to importing a plugin. However, a parser must be a separate module.

Reviewer 2: ```suggestion
In eslintrc files, importing a custom parser is similar to importing a plugin: you use a string to specify the name of the parser.
```

> In flat config files, import a custom parser as a module. Then assign it to the `languageOptions.parser` property of a configuration object.

Reviewer 2: ```suggestion
In flat config files, import a custom parser as a module then assign it to the `languageOptions.parser` property of a configuration object.
```

Author: ```suggestion
In flat config files, import a custom parser as a module, then assign it to the `languageOptions.parser` property of a configuration object.
```

if keeping 1 sentence should be a comma

                semi: ["warn", "always"]
            }
        },
        {
            files:["test/**/*"],
            rules: {
                "no-console": "off"
            }
        }
    ]
};
```

#### Flat Config Examples

Here is a configuration with the default glob pattern:

```javascript
// eslint.config.js

import js from "@eslint/js";

export default [
   js.configs.recommended, // Recommended config applied to all files
    // Override the recommended config
    {
        rules: {
           indent: ["error", 2],
           "no-unused-vars": "warn"
        }
        // ...other configuration
    }
];
```

An example configuration supporting multiple configs for different glob patterns:

```javascript
// eslint.config.js

import js from "@eslint/js";

export default [
    js.configs.recommended, // Recommended config applied to all files
    // File-pattern specific overrides
    {
        files: ["src/**/*", "test/**/*"],
        rules: {
                semi: ["warn", "always"]
        }
    },
    {
        files:["test/**/*"],
        rules: {
            "no-console": "off"
        }
    }
    // ...other configurations
];
```

### Configuring Language Options

> In eslintrc files, you configure various language options across the `env`, `globals` and `parserOptions` properties. Global variables for specific runtimes (e.g `document` for browser JavaScript and `process` for Node.js ) are configured with the `env` property.

Reviewer 2: ```suggestion
In eslintrc files, you configure various language options across the `env`, `globals` and `parserOptions` properties. Groups of global variables for specific runtimes (e.g `document` and `window` for browser JavaScript; `process` and `require` for Node.js ) are configured with the `env` property.
```

> In the flat config files, the `globals`, and `parserOptions` are consolidated under the `languageOptions` key. There is no longer the `env` property in `eslint.config.js`. Global variables for specific runtimes are imported from the [globals](https://www.npmjs.com/package/globals) npm package and included in the `globals` property. You can use the spread operator (`...`) to import all globals.

Reviewer 2: ```suggestion
In flat config files, the `globals`, and `parserOptions` are consolidated under the `languageOptions` key; the `env` property doesn't exist. Groups of global variables for specific runtimes are imported from the [globals](https://www.npmjs.com/package/globals) npm package and included in the `globals` property. You can use the spread operator (`...`) to import multiple globals at once.
```

#### eslintrc Example

```javascript
// .eslintrc

module.exports = {
    env: {
        browser: true,
    },
    globals: {
        myCustomGlobal: "readonly",
    },
    parserOptions: {
        ecmaVersion: 2022,
        sourceType: "module"
    }
    // ...other config
}
```

#### Flat Config Example

```javascript
// eslint.config.js

import globals from "globals";

export default [
    {
        languageOptions: {
            globals: {
                ...globals.browser,
                myCustomGlobal: "readonly"
            },
            parserOptions: {
                ecmaVersion: 2022,
                sourceType: "module"
            }
        }
        // ...other config
    }
];
```

### Predefined Configs

In eslintrc files, use the `extends` property to use predefined configs. ESLint comes with two predefined configs that you can access as strings:

* `"eslint:recommended"`: the rules recommended by ESLint
* `"eslint:all"`: all rules shipped with ESLint

You can also use the `extends` property to extend a custom config. Custom configs can either be paths to local config files or npm package names.

> In flat config files, predefined configs are imported from separate modules into flat confg files. The `recommended` and `all` rules configs are located in the [`@eslint/js`](https://www.npmjs.com/package/@eslint/js) package. You must import this package to use these configs:

Reviewer: ```suggestion
In flat config files, predefined configs are imported from separate modules into flat config files. The `recommended` and `all` rules configs are located in the [`@eslint/js`](https://www.npmjs.com/package/@eslint/js) package. You must import this package to use these configs:
```

```shell
npm install @eslint/js --save-dev
```

You can add each of these configs to the exported array or expose specific rules from them. You must import the modules for local config files and npm package configs with flat confg.

#### eslintrc Examples

Use built-in config:

```javascript
// .eslintrc.js

module.exports = {
    // ...other config
    extends: "eslint:recommended",
    rules: {
        semi: ["warn", "always"]
    },
    // ...other config
}
```

Use built-in config, local custom config, and custom config from an npm package:

```javascript
// .eslintrc.js

module.exports = {
    // ...other config
    extends: ["eslint:recommended", "./custom-config.js", "eslint-config-my-config"],
    rules: {
        semi: ["warn", "always"]
    },
    // ...other config
}
```

#### Flat Config Examples

Use recommended config:

```javascript
// eslint.config.js

import js from "@eslint/js";
import customConfig from "./custom-config.js";
import myConfig from "eslint-config-my-config";

export default [
    js.configs.recommended,
    customConfig,
    myConfig,
    {
        rules: {
            semi: ["warn", "always"]
        },
        // ...other config
    }
];
```

> Note that because you are just importing JavaScript modules, you can mutate the config objects before exporting them. For example, you might want to have a certain config object only apply to your test files:

Reviewer 2: ```suggestion
Note that because you are just importing JavaScript modules, you can mutate the config objects before ESLint uses them. For example, you might want to have a certain config object only apply to your test files:
```

To ignore files with flat config, you can use the `ignores` property in a config object. The `ignores` property accepts an array of glob patterns. Note that flat config glob patterns do _not_ match dot files (e.g. `.dotfile.js`).

#### eslintrc Examples

`.eslintignore` example:

```shell
# .eslintignore
temp.js
.config/*
# ...other ignored files
```

`ignorePatterns` example:

```javascript
// .eslintrc.js
{
    // ...other config
    ignorePatterns: ["temp.js", ".config/*"],
}
```

#### Flat Config Examples

```javascript
export default [
    {
        // ...other config
        ignores: ["temp.js", ".config/*"]
    }
];
```

### CLI Flag Changes

The following CLI flags are no longer supported with the flat config file format:

* `--rulesdir`
* `--ext`
* `--resolve-plugins-relative-to`

The flag `--no-eslintrc` has been replaced with `--no-config-lookup`.

### Additional Changes

The following changes have been made from the eslintrc to the flat config file format:

> * The `root` option no longer exists.

Reviewer 2: ```suggestion
* The `root` option no longer exists. (Flat config files act as if `root: true` is set.)
```

* The `files` option cannot be a single string anymore, it must be an array.
* The `sourceType` option now supports the new value `"commonjs"` (`.eslintrc` supports it too, but it was never documented).

> * You can configure `noInlineConfig` and `reportUnusedDisableDirectives` options under the setting `linterOptions`.

Reviewer 2: I think we could use a section with an example for this instead of it just being a bullet point here.

Author: note to self: still have to implement

Author: implemented

---
title: Configuration Migration Guide
eleventyNavigation:
    key: migration guide
    parent: configure
    title: Configuration Migration Guide
    order: 8
---

> This guide provides an overview of how you can migrate your ESLint configuration file from the `.eslintrc` format to the new configuration file format, `eslint.config.js`.

Reviewer 2: Because these config formats can exist outside of files, I think it's helpful to name them and then refer to the names throughout rather than the filenames.

```suggestion
This guide provides an overview of how you can migrate your ESLint configuration file from the eslintrc format (typically configured in `.eslintrc.js` or `.eslintrc.json` files) to the new flat config format (typically configured in an `eslint.config.js` file).
```

Author: will update throughout as well

> To learn more about the `eslint.config.js` file format, refer to [this blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/).

Reviewer 2: ```suggestion
To learn more about the flat config format, refer to [this blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/).
```

For reference information on these configuration formats, refer to the following documentation:

> * [`.eslintrc` configuration files](configuration-files)
> * [`eslint.config.js` configuration files](configuration-files-new)

Reviewer 2: ```suggestion
* [eslintrc configuration files](configuration-files)
* [flat configuration files](configuration-files-new)
```

## Start Using `eslint.config.js`

Starting with ESLint v9.0.0, the `eslint.config.js` file format will be the default configuration file format. Once ESLint v9.0.0 is released, you can start using the `eslint.config.js` file format without any additional configuration.

To use `eslint.config.js` with ESLint v8, place a `eslint.config.js` file in the root of your project **or** set an `ESLINT_USE_FLAT_CONFIG` environment variable to `true`.

## Key Differences between Configuration Formats

A few of the most notable differences between the `.eslintrc` and `eslint.config.js` formats are the following:

### Importing Plugins and Custom Parsers

> #### .eslintrc

Reviewer 2: I think it would be helpful to summarize the difference under each header and then show the examples without headers. Maybe something like:

> In eslintrc, plugins are loaded by specifying the name of the plugin in the `plugins` array; in flat config, you'll need to import the plugin object and then assign it to a property in the `plugins` object.

### Custom Parsers

#### .eslintrc

Importing a custom parser is very similar to importing a plugin. However, a parser must be a separate module.

```javascript
// .eslintrc.js

{
    // ...other config
    parser: "@babel/eslint-parser",
    // ...other config
}
```

#### eslint.config.js

To import a custom parser, import the parser as a module. Then assign it to the `languageOptions.parser` property of a configuration object:

```javascript
// eslint.config.js

import babelParser from "@babel/eslint-parser";

export default [
   {
       // ...other config
       languageOptions: {
           parser: babelParser
       }
       // ...other config
   }
];
```

### Glob-Based Configs

#### .eslintrc

By default, .`eslintrc` files lint all files (except those covered by `.gitignore`) in the directory in which they’re placed and its child directories:

```javascript
// .eslintrc.js

module.exports = {
    // ...other config
    rules: {
        semi: ["warn", "always"]
    }
};
```

If you want to have different configurations for different file glob patterns, you can specify them in the `overrides` property:

```javascript
// .eslintrc.js

module.exports = {
    // ...other config

>     extends: "eslint:recommended",

Reviewer 2: I'd remove this line as `extends` isn't in the previous config and doesn't really factor into the discussion.
```suggestion
```

```javascript
// eslint.config.js

import js from "@eslint/js";
import customTestConfig from "./custom-test-config.js";

export default [
    js.configs.recommended,
    {
        ...customTestConfig,
        files: ["**/*.test.js"],
    },
];
```

### Ignoring Files

#### .eslintrc

To make ESLint ignore files with `.eslintrc`, you can create a separate `.eslintignore` file in the root of your project. The `.eslintignore` file uses the same glob pattern syntax as `.gitignore` files.

```shell
# .eslintignore
temp.js
.config/*
# ...other ignored files
```

Alternatively, you can use an `ignorePatterns` property in your `.eslintrc` file:

```javascript
// .eslintrc.js
{
    // ...other config
    ignorePatterns: ["temp.js", ".config/*"],
}
```

#### eslint.config.js

To ignore files with `eslint.config.js`, you can use the `ignores` property in a config object:

```javascript
export default [
    {
        // ...other config
        ignores: ["temp.js", ".config/*"]
    }
];
```

> Also, dotfiles (e.g. `.dotfile.js`) are no longer ignored by default with `eslint.config.js`.

Reviewer 3: re the dotfiles: in a later tsc meeting, we've decided to keep it as-is:
https://github.com/eslint/tsc-meetings/pull/378/files#diff-86bfb8d4387c1278423e2710f8108a542396d2bc131cce91479a7b3fbfec23e5R27

Reviewer: > re the dotfiles: in a later tsc meeting, we've decided to keep it as-is: https://github.com/eslint/tsc-meetings/pull/378/files#diff-86bfb8d4387c1278423e2710f8108a542396d2bc131cce91479a7b3fbfec23e5R27

I think that resolution is about glob patterns, so for example `"ignores": "src/**"` shall not ignore files whose name starts with a dot. If `ignores` pattern are not specified, by default, dot-files are linted as well.

Author: i'm not sure if i fully understand but will refactor and y'all can comment on if it needs further refinement.

### CLI Flag Changes

The following CLI flags are no longer supported with the `eslint.config.js` file format:

* `--rulesdir`
* `--ext`
* `--resolve-plugins-relative-to`

The flag `--no-eslintrc` has been replaced with `--no-config-lookup`.

### Additional Changes

The following changes have been made from the `.eslintrc` to the `eslint.config.js` file format:

* The `root` option no longer exists.
* The `files` option cannot be a single string anymore, it must be an array.
* The `sourceType` option now supports the new value `"commonjs"` (`.eslintrc` supports it too, but it was never documented).
* You can configure `noInlineConfig` and `reportUnusedDisableDirectives` options under the setting `linterOptions`.

## Things That Haven’t Changed between Configuration File Formats

While all the above mentioned features have changed from `.eslintrc` to `eslint.config.js` configurations, the following has stayed the same:

> * Syntax for adding rules

Author: yeah i meant 'adding rules to the configuration file', so i think saying 'Rule configuration' expresses that more concisely.

* Processors

> * All functionality. Just the way to configure it has changed.

Reviewer 2: I'm not sure what "all functionality" means here. I think we can remove this.

Author: by this, i mean "how ESLint lints files and reports errors". just the way to set up the configuration has changed. 

i'll play w the phrasing a bit here.

}
```

To use the same configs in flat config, you would do the following:

```javascript
// eslint.config.js

import js from "@eslint/js";
import customConfig from "./custom-config.js";
import myConfig from "eslint-config-my-config";

export default [
    js.configs.recommended,
    customConfig,
    myConfig,
    {
        rules: {
            semi: ["warn", "always"]
        },
        // ...other config
    }
];
```

Note that because you are just importing JavaScript modules, you can mutate the config objects before ESLint uses them. For example, you might want to have a certain config object only apply to your test files:

```javascript
// eslint.config.js

import js from "@eslint/js";
import customTestConfig from "./custom-test-config.js";

export default [
    js.configs.recommended,
    {
        ...customTestConfig,
        files: ["**/*.test.js"],
    },
];
```

### Ignoring Files

With eslintrc, you can make ESLint ignore files by creating a separate `.eslintignore` file in the root of your project. The `.eslintignore` file uses the same glob pattern syntax as `.gitignore` files. Alternatively, you can use an `ignorePatterns` property in your eslintrc file.

> To ignore files with flat config, you can use the `ignores` property in a config object. The `ignores` property accepts an array of glob patterns. Note that flat config glob patterns do _not_ match dot files (e.g. `.dotfile.js`).

Reviewer 5: Hi everyone,
Maybe we should clarify that [old .eslintignore no longer works with flat config](https://github.com/eslint/eslint/discussions/17342#discussioncomment-6359326)?
I'm not fully sure that it's understandable with the current ignore docs info
Just a small notion:

```suggestion
To ignore files with flat config, you can use the `ignores` property in a config object. The `ignores` property accepts an array of glob patterns. Note that flat config glob patterns do _not_ match dot files (e.g. `.dotfile.js`). Please note that since v8.25.0 the [new flat config system no longer supports .eslintignore files](https://eslint.org/blog/2022/10/eslint-v8.25.0-released/#highlights).
```

Reviewer 2: Probably a good idea, though I'd prefer this:

```suggestion
To ignore files with flat config, you can use the `ignores` property in a config object. The `ignores` property accepts an array of glob patterns. Note that flat config glob patterns do _not_ match dot files (e.g. `*.js` won't match `.dotfile.js`). Flat config does not support loading ignore patterns from `.eslintignore` files, so you'll need to migrate those patterns directly into flat config.
```

The move away from `.eslintignore` was fairly early in the flat config lifecycle, so I don't think we need to mark it in time.

Reviewer 5: > Probably a good idea, though I'd prefer this:
> 
> The move away from `.eslintignore` was fairly early in the flat config lifecycle, so I don't think we need to mark it in time.

Maybe you're right)
I started using flat config right after the release. And during updates, I had to figure out why .eslintignore no longer works for the flat config. And the only mention where I found it was in this blog post. So I think It's a little confusing. But I agree that this is a rare situation. And your correction fits

Author: adding your suggestion @nzakas

For example, here's an eslintrc file with linter options enabled:

```javascript
// .eslintrc.js

module.exports = {
    // ...other config
    noInlineConfig: true,
    reportUnusedDisableDirectives: true
}
```

Here's the same options in flat config:

```javascript
// eslint.config.js

export default [
    {
        // ...other config
        linterOptions: {
            noInlineConfig: true,
            reportUnusedDisableDirectives: true
        }
    }
];
```

### CLI Flag Changes

The following CLI flags are no longer supported with the flat config file format:

* `--rulesdir`
* `--ext`
* `--resolve-plugins-relative-to`

The flag `--no-eslintrc` has been replaced with `--no-config-lookup`.

### Additional Changes

The following changes have been made from the eslintrc to the flat config file format:

* The `root` option no longer exists. (Flat config files act as if `root: true` is set.)
* The `files` option cannot be a single string anymore, it must be an array.
* The `sourceType` option now supports the new value `"commonjs"` (`.eslintrc` supports it too, but it was never documented).

## TypeScript Types for Flat Config Files

> You can see the TypeScript types for the eslint.config.js file format in the DefinitelyTyped project. The interface for the objects in the config’s array is called the `FlatConfig`.

Reviewer: ```suggestion
You can see the TypeScript types for the `eslint.config.js` file format in the DefinitelyTyped project. The interface for the objects in the config’s array is called the `FlatConfig`.
```

Shall we use backticks here for consistency with the other filenames in the document?

Author: ```suggestion
You can see the TypeScript types for the flat config file format in the DefinitelyTyped project. The interface for the objects in the config’s array is called the `FlatConfig`.
```

You can view the type definitions in the [DefinitelyTyped repository on Github](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/eslint/index.d.ts).

## Further Reading

> * [Overview of the eslint.config.js file format blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/)

Reviewer: ```suggestion
* [Overview of the `eslint.config.js` file format blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/)
```

And here?

Author: ```suggestion
* [Overview of the flat config file format blog post](https://eslint.org/blog/2022/08/new-config-system-part-2/)
```

### Importing Plugins and Custom Parsers

> Eslintrc files use string-based import system with the `plugins` property to load plugins and and with the `extends` to load external configurations.

Reviewer 2: ```suggestion
Eslintrc files use string-based import system inside the `plugins` property to load plugins and inside the `extends` property to load external configurations.
```

};
```

> For the flat config, here is a configuration with the default glob pattern:

Reviewer 2: ```suggestion
For flat config, here is a configuration with the default glob pattern:
```

### Configuring Language Options

In eslintrc files, you configure various language options across the `env`, `globals` and `parserOptions` properties. Groups of global variables for specific runtimes (e.g `document` and `window` for browser JavaScript; `process` and `require` for Node.js ) are configured with the `env` property.

In flat config files, the `globals`, and `parserOptions` are consolidated under the `languageOptions` key; the `env` property doesn't exist. Groups of global variables for specific runtimes are imported from the [globals](https://www.npmjs.com/package/globals) npm package and included in the `globals` property. You can use the spread operator (`...`) to import multiple globals at once.

> #### eslintrc Example

Reviewer 2: ```suggestion
```

}
```

> Here's the same files ignored in flat config:

Reviewer 2: ```suggestion
Here are the same files ignore patterns in flat config:
```
